### ISSUE TITLE
fetch fails given cloned request object with form data.

### ISSUE THREAD

**Issue by pvormittag** [denoland/deno#16570]
I'm running into `fetch()` throwing "Unreachable" from https://github.com/denoland/deno/blob/v1.26.1/ext/fetch/26_fetch.js#L169. This will happen any time `fetch()` is given a cloned request object that has form data. The issue does not occur when `body` is a `string`, but it may also fail with any of the other `BodyInit` types (i.e `Blob`, `BufferSource`, etc), which I haven't tested.

I can reproduce the issue by slightly altering the same example from https://github.com/denoland/deno/issues/10002.

```js
// Succeeds

await fetch(new Request('http://example.com').clone());
await fetch(new Request('http://example.com', { method: 'POST', body: JSON.stringify({foo:'bar'})}).clone());

// Throws "TypeError: Unreachable"

const formData = new FormData();
formData.append('foo', 'bar');
await fetch(new Request('http://example.com', { method: 'POST', body: formData }).clone());
```

Tested with:

deno 1.26.1 (release, aarch64-apple-darwin)
v8 10.7.193.3
typescript 4.8.3

deno 1.27.1 (release, aarch64-apple-darwin)
v8 10.8.168.4
typescript 4.8.3

**Comment by 0f-0b** [comment]
This happens because `InnerBody.clone` would fail to serialize `source` when it is a `Blob` (#12067),

https://github.com/denoland/deno/blob/cbb3f854332c348bb253e1284f7dcd7287bdf28d/ext/fetch/22_body.js#L191

but this line expects `source` to be properly cloned.

https://github.com/denoland/deno/blob/cbb3f854332c348bb253e1284f7dcd7287bdf28d/ext/fetch/26_fetch.js#L157

**Comment by pvormittag** [comment]
Thank you for the insight @0f-0b! Is there a potential work-around for this issue from user land code, or does this require a patch to Deno core? I presume the latter, but for my current case I'm able to except a work-around, so any further insight would be appreciated!

**Comment by 0f-0b** [comment]
Sorry for the late reply. I encountered this problem when using [ky](https://github.com/sindresorhus/ky) which clones every request but allows to modify the request just before it is cloned. My workaround is to collect the request body into an `ArrayBuffer` so that `InnerBody.source` would be a serializable `Uint8Array`.

```ts
// request.ts
export async function toClonableRequest(req: Request): Promise<Request> {
  if (!req.body) {
    return req;
  }
  return new Request(req.url, {
    body: await req.arrayBuffer(),
    cache: req.cache,
    credentials: req.credentials,
    headers: req.headers,
    integrity: req.integrity,
    keepalive: req.keepalive,
    method: req.method,
    mode: req.mode,
    redirect: req.redirect,
    referrer: req.referrer,
    referrerPolicy: req.referrerPolicy,
    signal: req.signal,
  });
}
```

```ts
import { ky } from "./deps.ts";
import { toClonableRequest } from "./request.ts";

const form = new FormData();
form.append("foo", "bar");

await ky.post("http://example.com/", {
  body: form,
  hooks: { beforeRequest: [toClonableRequest] },
});

// roughly equivalent to
const req = new Request("http://example.com/", {
  body: form,
  method: "POST",
  signal: AbortSignal.timeout(10000),
});
await fetch((await toClonableRequest(req)).clone());
```

---

To actually fix this issue in Deno, I suppose it would suffice to just remove the unnecessary serialization / deserialization, since `InnerBody.source` is immutable anyway.

**Comment by pvormittag** [comment]
> My workaround is to collect the request body into an `ArrayBuffer` so that `InnerBody.source` would be a serializable `Uint8Array`.

That makes sense and I was working down a similar path for a hopeful workaround, so thank you for sharing! I'm sure others will run into this as well.

**Comment by 0f-0b** [comment]
This seems to have been fixed in #20217.

**Comment by lucacasonato** [comment]
This is fixed
